src/core/quantum_gate.jl

```diff
@@ -174,13 +117,112 @@ function apply_gate!(state::Ket, gate::AbstractGate)
     operator=get_operator(gate,type_in_ket)
 
     apply_operator!(state,operator,connected_qubits)
+end
+
+# specialization for single target dense gate (size N=2, for N=2^target_count)
+function apply_operator!(
+    state::Ket,
+    operator::DenseOperator{2},
+    connected_qubit::Vector{<:Integer})
+
+    qubit_count = Int(log2(length(state)))
+
+    dim=2^qubit_count
+
+    # the bitwise implementation assumes target numbering starting at 0,
+    # with first qubit on the rightmost side
+    target_qubit_index=qubit_count-connected_qubit[1] 
+    
+    loop_dim = div(dim,2);
+    mask = (UInt64(1) << target_qubit_index);
+    mask_low = mask - 1;
+    mask_high = ~mask_low;
+    
+    matrix=operator.data
+
+    for state_index in UnitRange{UInt64}(UInt64(0),UInt64(loop_dim-1))
+        basis_0 =(state_index & mask_low) + ((state_index & mask_high) << 1)
+        basis_1 = basis_0 + mask
+
+        # fetch values
+        @inbounds cval_0 = state.data[basis_0+1];
+        @inbounds cval_1 = state.data[basis_1+1];
+
+        @inbounds state.data[basis_0+1] = matrix[1] * cval_0 + matrix[3] * cval_1;
+        @inbounds state.data[basis_1+1] = matrix[2] * cval_0 + matrix[4] * cval_1;
+    end
+end
+
+# specialization for multiple target dense gate (size N>=4, for N=2^target_count)
+function apply_operator!(
```

Comment:
Are you using any specific reference for this implementation? If so, please cite it here.